File: Yandex/Practicum/Algorightms_and_Data_Structures/8_sprint/E._Вставка_строк/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(N + M)
    Memory Complexity O(M)
    """
    s: str = sys.stdin.readline().rstrip()
    n: int = int(sys.stdin.readline().rstrip())
    gifts: dict[int, str] = dict()
    for _ in range(n):
        t, k = sys.stdin.readline().rstrip().split()
        gifts[int(k)] = t
    for i in range(len(s)):
        if i in gifts:
            sys.stdout.write(gifts[i])
        sys.stdout.write(s[i])
    if len(s) in gifts:
        sys.stdout.write(gifts[len(s)])
    sys.stdout.write('\n')


if __name__ == '__main__':
    solution()

File: Yandex/Practicum/Algorightms_and_Data_Structures/8_sprint/E._Вставка_строк/solution.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <string>


void solution() {
    /*
    Time Complexity O(N + M)
    Memory Complexity O(M)
    */
    std::string s;
    std::cin >> s;
    int n;
    std::scanf("%d", &n);
    std::unordered_map<int, std::string> gifts;
    for (int i = 0; i < n; ++i) {
        std::string t;
        std::cin >> t;
        int k;
        std::scanf("%d", &k);
        gifts[k] = t;
    }

    for (int i = 0; i < s.size(); ++i) {
        if (gifts.count(i)) std::cout << gifts[i];
        std::cout << s[i];
    }
    if (gifts.count(s.size())) {
        std::cout << gifts[s.size()];
    }
    std::cout << "\n";
}


int main() {
    solution();
}